Spec adding span links for instrumentation of messaging systems (#616)
elastic/apm: specs/agents/tracing-instrumentation-messaging.md

### Trace Context

This section applies to messaging systems that support [message metadata](#message-metadata).
The instrumentation of message reception SHOULD check message metadata for the
presence of [Trace Context](https://www.w3.org/TR/trace-context/).

When single message reception is captured as a `messaging` transaction,

> if Trace Context is present, it SHOULD be propagated to the `messaging` transaction

Reviewer: ```suggestion
and a Trace Context is present, it SHOULD be used as the parent of the `messaging` transaction
```

to continue the [distributed trace](tracing-distributed-tracing.md).

Otherwise (a single message being captured as a `messaging` span, or a batch
of messages is processed in a single `messaging` transaction or span), a
[span link](span-links.md) SHOULD be added for each message with Trace Context.

> 

Reviewer: ```suggestion
This includes the case where the size of the batch of received messages is one.

```

| Central config | `true` |

### Message metadata

To support distributed tracing with automatic instrumentation, the messaging
system must provide a mechanism to add metadata/properties/attributes to
individual messages, akin to HTTP headers. If an APM agent supports
trace-context for a given messaging system, it MUST use the following mechanisms
so that cross-language tracing works:

| Messaging system       | Mechanism |
| ---------------------- | --------- |
| Azure Queue            | ??? |
| Azure Service Bus      | ??? |

> | Java Messaging Service | ??? |

Reviewer: ```suggestion
| Java Messaging Service | [Message Properties](https://docs.oracle.com/javaee/7/api/javax/jms/Message.html) |
```

Author: Thanks. Commit 13dc897

> | Apache Kafka           | TODO: is the Java APM agent using [Kafka Record headers](https://cwiki.apache.org/confluence/display/KAFKA/KIP-82%2B-%2BAdd%2BRecord%2BHeaders)? |

Author: Thanks. Commit 13dc897

> | RabbitMQ               | TODO: are [Message Attributes](https://www.rabbitmq.com/tutorials/amqp-concepts.html#messages) feasible? |

Reviewer: Yes. See also https://github.com/elastic/apm-agent-java/blob/main/apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-plugin/src/main/java/co/elastic/apm/agent/rabbitmq/header/RabbitMQTextHeaderGetter.java.

In the Java API, message attributes are referred to as `AMQP.BasicProperties` but it seems to be the same thing.

Author: Thanks. Commit 13dc897

| Central config | `true` |

### Message metadata

To support distributed tracing with automatic instrumentation, the messaging
system must provide a mechanism to add metadata/properties/attributes to
individual messages, akin to HTTP headers. If an APM agent supports
trace-context for a given messaging system, it MUST use the following mechanisms
so that cross-language tracing works:

| Messaging system       | Mechanism |
| ---------------------- | --------- |

> | Azure Queue            | Unknown |

Reviewer 2: IIUC, Azure Queue Messages do not have a means to transport metadata / headers / attributes.

> | Azure Service Bus      | Unknown |

Reviewer 2: For Azure Service Bus it would be this one, I guess: https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-end-to-end-tracing?tabs=net-standard-sdk-2

Author: Thanks. I added a note in https://github.com/elastic/apm/pull/616/commits/6f73489626bbf8d039e9b9c6f345625e14b25005
